fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/CascadesPlanner.java

```diff
@@ -565,35 +611,61 @@ public String toString() {
     /**
      * Explore Group Task.
      * <br>
-     * Simplified enqueue/execute overview:
+     * Simplified push/execute overview:
      * <br>
      * {@link ExploreGroup}
-     *     enqueues
+     *     pushes
      *         {@link ExploreExpression} for each group member
      *     sets explored to {@code true}
      */
     private class ExploreGroup implements Task {
         @Nonnull
-        private final PlanContext context;
+        private final PlannerPhase plannerPhase;
         @Nonnull
         private final Reference group;
-        @Nonnull
-        private final EvaluationContext evaluationContext;
 
-        public ExploreGroup(@Nonnull PlanContext context,
-                            @Nonnull Reference ref,
-                            @Nonnull final EvaluationContext evaluationContext) {
-            this.context = context;
+        public ExploreGroup(@Nonnull final PlannerPhase plannerPhase,
+                            @Nonnull final Reference ref) {
+            this.plannerPhase = plannerPhase;
             this.group = ref;
-            this.evaluationContext = evaluationContext;
+        }
+
+        @Nonnull
+        @Override
+        public PlannerPhase getPlannerPhase() {
+            return plannerPhase;
         }
 
         @Override
         public void execute() {
+            if (plannerPhase.getTargetStage().ordinal() < group.getPlannerStage().ordinal()) {
```

Comment:
This will be changed to:

```
if targetStage of plannerPhase < group's stage:
   unshare sub DAG underneath group;
   for all planner phases p from plannerPhase until targetStage of p == groupStage:
      explore and optimize in p;
if targetStage of plannerPhase > group's stage:
   advancePlannerStage();
else:
   do nothing;

Reply:
Currently we just bail if we detect `targetStage of plannerPhase < group's stage`